src/VisualStudio/Core/Def/Implementation/ProjectSystem/MetadataReferences/FileWatchedPortableExecutableReferenceFactory.cs

```diff
@@ -0,0 +1,134 @@
+{
+    [Export]
+    internal sealed class FileWatchedPortableExecutableReferenceFactory
+    {
+        private readonly object _gate = new object();
+
+        /// <summary>
+        /// This right now acquires the entire VisualStudioWorkspace because right now the production
+        /// of metadata references depends on other workspace services. See the comments on
+        /// <see cref="VisualStudioMetadataReferenceManagerFactory"/> that this strictly shouldn't be necessary
+        /// but for now is quite the tangle to fix.
+        /// </summary>
+        private readonly Lazy<VisualStudioWorkspace> _visualStudioWorkspace;
+
+        /// <summary>
+        /// A file change context used to watch metadata references.
+        /// </summary>
+        private readonly FileChangeWatcher.IContext _fileReferenceChangeContext;
+
+        /// <summary>
+        /// File watching tokens from <see cref="_fileReferenceChangeContext"/> that are watching metadata references. These are only created once we are actually applying a batch because
+        /// we don't determine until the batch is applied if the file reference will actually be a file reference or it'll be a converted project reference.
+        /// </summary>
+        private readonly Dictionary<PortableExecutableReference, FileChangeWatcher.IFileWatchingToken> _metadataReferenceFileWatchingTokens = new Dictionary<PortableExecutableReference, FileChangeWatcher.IFileWatchingToken>();
+
+        /// <summary>
+        /// <see cref="CancellationTokenSource"/>s for in-flight refreshing of metadata references. When we see a file change, we wait a bit before trying to actually
+        /// update the workspace. We need cancellation tokens for those so we can cancel them either when a flurry of events come in (so we only do the delay after the last
+        /// modification), or when we know the project is going away entirely.
+        /// </summary>
+        private readonly Dictionary<string, CancellationTokenSource> _metadataReferenceRefreshCancellationTokenSources = new Dictionary<string, CancellationTokenSource>();
+
+        [ImportingConstructor]
+        public FileWatchedPortableExecutableReferenceFactory(
+            Lazy<VisualStudioWorkspace> visualStudioWorkspace,
+            FileChangeWatcherProvider fileChangeWatcherProvider)
+        {
+            _visualStudioWorkspace = visualStudioWorkspace;
+
+            // TODO: set this to watch the NuGet directory or the reference assemblies directory; since those change rarely and most references
+            // will come from them, we can avoid creating a bunch of explicit file watchers.
+            _fileReferenceChangeContext = fileChangeWatcherProvider.Watcher.CreateContext();
+            _fileReferenceChangeContext.FileChanged += FileReferenceChangeContext_FileChanged;
+        }
+
+        public event EventHandler<string> ReferenceChanged;
```

Comment:
>EventHandler [](start = 21, length = 12)

Curious, why are we using EventHandler instead of just `Action<string>`. We never seem to use `object sender` parameter.

Reply:
Just following framework convention. And we don't use it now, but if this ever goes public, better to do it than regret it.